operator/controllers/helmrelease_controller.go

```diff
@@ -83,19 +97,28 @@ func (r *HelmReleaseReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-func (r *HelmReleaseReconciler) onDelete(ctx context.Context, name types.NamespacedName) error {
-	// Update HelmRelease Status to 'PENDING_DELETE' -- but the HelmRelease doesn't exist anymore :(
+func (r *HelmReleaseReconciler) setFinalizer(ctx context.Context, rls shipitv1beta1.HelmRelease) error {
+	finalizers := rls.GetFinalizers()
+	rls.SetFinalizers(append(finalizers, HelmReleaseFinalizer))
+
+	return r.Update(ctx, &rls)
+}
+
+func (r *HelmReleaseReconciler) onDelete(ctx context.Context, rls shipitv1beta1.HelmRelease) error {
+	// Update HelmRelease Status to 'DELETING'
 	// Delete the release with helm
 	return errNotImplemented
 }
 
-func (r *HelmReleaseReconciler) onUpdate(ctx context.Context, rls shipitv1beta1.HelmRelease) error {
-	// Update HelmRelease Status to 'PENDING_INSTALL' or 'PENDING_UPGRADE'
-	// Attempt the install/upgrade with helm
-	// If it succeeded, set Status to 'DEPLOYED'
-	// Else, set Status to 'PENDING_ROLLBACK'
-	// Attempt the rollback with helm
-	// If it succeeded, set Status to 'DEPLOYED'
-	// Else, set Status to 'FAILED'
-	return errNotImplemented
+func (r *HelmReleaseReconciler) onUpdate(ctx context.Context, rls shipitv1beta1.HelmRelease) (ctrl.Result, error) {
```

Comment:
This returns a `ctrl.Result`, whereas `onDelete` does not, to allow simply re-queuing a delayed reconciliation request after an install/update. This would allow us to implement the desired async behaviour without having to maintain ownership references on the resources created by helm, although it's less robust and efficient overall. Not totally sure which way I'd prefer to go yet